File: VB/DXPivotGrid_CustomFilterPopup/CustomFilterEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpf.Editors
Imports System.Windows
Imports DevExpress.Xpf.PivotGrid
Imports System.Windows.Input
Imports DevExpress.Mvvm
Imports DevExpress.Xpf.Core.Native
Imports System.Windows.Controls
Imports System.Threading
Imports System.Windows.Threading
Imports DevExpress.Mvvm.UI
Imports DevExpress.Xpf.PivotGrid.Internal
Imports DevExpress.Xpf.Grid.LookUp

Namespace DXPivotGrid_CustomFilterPopup

	Friend Class PivotGridCustomFilter
		Inherits LookUpEdit

		Public Sub New()
			AddHandler Me.PopupOpening, AddressOf PivotGridRadioFilter_PopupOpening
			AddHandler Me.PopupClosed, AddressOf PivotGridRadioFilter_PopupClosed
		End Sub
		Public Property PivotField() As PivotGridField
			Get
				Return DirectCast(GetValue(PivotFieldProperty), PivotGridField)
			End Get
			Set(ByVal value As PivotGridField)
				SetValue(PivotFieldProperty, value)
			End Set
		End Property
		Public Shared ReadOnly PivotFieldProperty As DependencyProperty = DependencyProperty.Register("PivotField", GetType(PivotGridField), GetType(PivotGridCustomFilter), New UIPropertyMetadata(Nothing))

		Private Sub PivotGridRadioFilter_PopupOpening(ByVal sender As Object, ByVal e As OpenPopupEventArgs)
			If PivotField Is Nothing Then
				e.Cancel = True
				e.Handled = True
				Return
			End If
			Me.BeginDataUpdate()

			Dim uniqueItems = PivotField.GetUniqueValues().Select(Function(v) New PivotValueInfo() With {
				.Value = v,
				.DisplayText = PivotField.GetDisplayText(v)
			}).ToArray()
'INSTANT VB NOTE: The variable editValue was renamed since Visual Basic does not handle local variables named the same as class members well:
			Dim editValue_Renamed = PivotField.FilterValues.ValuesIncluded
			ItemsSource = uniqueItems

			Me.EditValue = editValue_Renamed
			Me.EndDataUpdate()
		End Sub

		Private Sub PivotGridRadioFilter_PopupClosed(ByVal sender As Object, ByVal e As ClosePopupEventArgs)
			Dim pivot As PivotGridControl = TryCast(PivotField.Parent, PivotGridControl)
			If e.CloseMode = DevExpress.Xpf.Editors.PopupCloseMode.Cancel OrElse PivotField Is Nothing OrElse SelectedItems Is Nothing OrElse pivot Is Nothing Then
				Return
			End If
			Dim values = SelectedItems.OfType(Of PivotValueInfo)().Select(Function(v) v.Value).ToArray()
			If PivotField.FilterValues.FilterType = FieldFilterType.Included Then
				PivotField.FilterValues.ValuesIncluded = values
			Else
				PivotField.FilterValues.ValuesExcluded = PivotField.GetUniqueValues().Except(values).ToArray()
			End If
		End Sub
	End Class

	Public Class PivotValueInfo
		Public Property Value() As Object
		Public Property DisplayText() As String
	End Class

End Namespace

File: CS/DXPivotGrid_CustomFilterPopup/CustomFilterEdit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpf.Editors;
using System.Windows;
using DevExpress.Xpf.PivotGrid;
using System.Windows.Input;
using DevExpress.Mvvm;
using DevExpress.Xpf.Core.Native;
using System.Windows.Controls;
using System.Threading;
using System.Windows.Threading;
using DevExpress.Mvvm.UI;
using DevExpress.Xpf.PivotGrid.Internal;
using DevExpress.Xpf.Grid.LookUp;

namespace DXPivotGrid_CustomFilterPopup {

    class PivotGridCustomFilter : LookUpEdit {
        public PivotGridCustomFilter() {
            PopupOpening += new OpenPopupEventHandler(PivotGridRadioFilter_PopupOpening);
            PopupClosed += new ClosePopupEventHandler(PivotGridRadioFilter_PopupClosed);
        }
        public PivotGridField PivotField {
            get { return (PivotGridField)GetValue(PivotFieldProperty); }
            set { SetValue(PivotFieldProperty, value); }
        }
        public static readonly DependencyProperty PivotFieldProperty =
            DependencyProperty.Register("PivotField", typeof(PivotGridField), typeof(PivotGridCustomFilter), new UIPropertyMetadata(null));

        void PivotGridRadioFilter_PopupOpening(object sender, OpenPopupEventArgs e) {
            if (PivotField == null) {
                e.Cancel = true;
                e.Handled = true;
                return;
            }
            this.BeginDataUpdate();
            
            var uniqueItems = PivotField.GetUniqueValues().Select(v => new PivotValueInfo() { Value = v, DisplayText = PivotField.GetDisplayText(v) }).ToArray();
            var editValue = PivotField.FilterValues.ValuesIncluded;
            ItemsSource = uniqueItems;
            
            this.EditValue = editValue;
            this.EndDataUpdate();
        }

        void PivotGridRadioFilter_PopupClosed(object sender, ClosePopupEventArgs e) {
            PivotGridControl pivot = PivotField.Parent as PivotGridControl;
            if (e.CloseMode == DevExpress.Xpf.Editors.PopupCloseMode.Cancel || PivotField == null || SelectedItems == null || pivot == null)
                return;
            var values = SelectedItems.OfType<PivotValueInfo>().Select(v => v.Value).ToArray();
            if (PivotField.FilterValues.FilterType == FieldFilterType.Included)
                PivotField.FilterValues.ValuesIncluded = values;
            else
                PivotField.FilterValues.ValuesExcluded = PivotField.GetUniqueValues().Except(values).ToArray();
        }
    }

    public class PivotValueInfo {
        public object Value { get; set; }
        public string DisplayText { get; set; }
    }

}
